packages/polywrap-uri-resolvers/polywrap_uri_resolvers/abc/resolver_with_history.py

```diff
@@ -0,0 +1,25 @@
+from abc import ABC, abstractmethod
+
+from polywrap_core import IUriResolver, Uri, IUriResolutionContext, UriPackageOrWrapper, Client, IUriResolutionStep
+from polywrap_result import Result
+
+
+class ResolverWithHistory(IUriResolver, ABC):
+    async def try_resolve_uri(self, uri: Uri, client: Client, resolution_context: IUriResolutionContext):
+        result = await self._try_resolve_uri(uri, client, resolution_context)
+        step = IUriResolutionStep(
+            source_uri=uri, 
+            result=result, 
+            description=self.get_step_description()
+        )
+        resolution_context.track_step(step)
+
+        return result
+
+    @abstractmethod
+    def get_step_description(self) -> str:
+        pass
+
+    @abstractmethod
+    async def _try_resolve_uri(self, uri: Uri, client: Client, resolution_context: IUriResolutionContext) -> Result["UriPackageOrWrapper"]:
```

Comment:
Why is this an abstract method? This should be optional and upto implementation of interface top define it or not.

Reply:
NVM I got this